config/encrypt_file_test.go

```diff
@@ -0,0 +1,87 @@
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/httpclient"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/httpclient/httpclientfakes"
+)
+
+var _ = Describe("Encrypt", func() {
+
+	const (
+		serviceRegistryInstance = "some-service-registry"
+		plainText               = "plain-text"
+		fileToEncrypt           = "file-to-encrypt.txt"
+		errorText               = "to err is human"
+		accessToken             = "access-token"
+		bearerAccessToken       = "bearer " + accessToken
+		serviceURI              = "service-uri/"
+		encryptURI              = "service-uri/encrypt"
+		cipherText              = "cipher-text"
+	)
+
+	var (
+		fakeCliConnection     *pluginfakes.FakeCliConnection
+		fakeAuthClient        *httpclientfakes.FakeAuthenticatedClient
+		fakeResolver          func(cliConnection plugin.CliConnection, serviceInstanceName string, accessToken string, authClient httpclient.AuthenticatedClient) (string, error)
+		resolverAccessToken   string
+		testError             error
+		postResponse          io.ReadCloser
+		postStatusCode        int
+		postErr               error
+		output                string
+		err                   error
+		accessTokenURI        string
+		fakeResolverCallCount int
+		fakeResolverError     error
+	)
+
+	BeforeEach(func() {
+		fakeCliConnection = &pluginfakes.FakeCliConnection{}
+		fakeAuthClient = &httpclientfakes.FakeAuthenticatedClient{}
+		testError = errors.New(errorText)
+		postResponse, postStatusCode, postErr = ioutil.NopCloser(bytes.NewBufferString(cipherText)), http.StatusOK, nil
+		accessTokenURI = "access-token-uri"
+		fakeCliConnection.AccessTokenReturns(bearerAccessToken, nil)
+		fakeResolverCallCount = 0
+		fakeResolver = func(cliConnection plugin.CliConnection, serviceInstanceName string, accessToken string, authClient httpclient.AuthenticatedClient) (string, error) {
+			fakeResolverCallCount++
+			resolverAccessToken = accessToken
+			return serviceURI, fakeResolverError
+		}
+		fakeResolverError = nil
+	})
+
+	JustBeforeEach(func() {
+		fakeAuthClient.DoAuthenticatedPostReturns(postResponse, postStatusCode, postErr)
+	})
+
+	It("should call the config server's /encrypt endpoint with content from a file", func() {
+		output, err = config.EncryptWithResolver(fakeCliConnection, serviceRegistryInstance, "", fileToEncrypt, fakeAuthClient, fakeResolver)
+
+		Expect(fakeAuthClient.DoAuthenticatedPostCallCount()).Should(Equal(1))
+		url, bodyType, body, token := fakeAuthClient.DoAuthenticatedPostArgsForCall(0)
+		Expect(url).To(Equal(encryptURI))
+		Expect(bodyType).To(Equal("text/plain"))
+		Expect(body).To(Equal("Hello\nWorld\n"))
+		Expect(token).To(Equal(accessToken))
+	})
+
+	It("should fail when given a non-existen", func() {
+		output, err = config.EncryptWithResolver(fakeCliConnection, serviceRegistryInstance, "", "bogus.txt", fakeAuthClient, fakeResolver)
+
+		Expect(fakeAuthClient.DoAuthenticatedPostCallCount()).Should(Equal(0))
+	})
+
```

Comment:
I'd like to see tests for other cases of the file path, e.g. `../file.txt` pointing to a valid file, `/absolute/path/file.txt` pointing to a valid file, `somedir` pointing to a directory to make sure the diagnostics are suitable.

Reply:
Is there a way to mock the file reading? I ask because while these cases are interesting (and, in fact, I've confirmed that they work manually, because originally I was testing with a file in my home directory and I was giving an absolute path to it), I'm not sure how to test with a relative path so that the relative path will be available no matter where the tests are run (CI vs. local dev).

I suppose, however, I could add tests for `../somefile.txt` and for a `somedir`. It'd be cleaner, however, if I could mock the file loading so that I won't have to litter the project with test files. It felt awkward enough dropping the one test file alongside the Go code.

The only thing I can think of is to wrap the actual file-loading code with something that itself could be mocked. At that point, I'm not sure what it is that I'm testing...how I react to bogus files? I certainly wouldn't want write tests that only test the wrapper or the mock.

Reply:
I usually construct a suitable set of directories and files for the test in a temporary directory. Take a look at that kind of thing [here](https://github.com/cf-guardian/guardian/blob/80736895e96f5cf4051554ee71b71155cc795812/kernel/fileutils/fileutils_test.go) (unfortunately before I started using ginkgo).

With a bit in ingenuity, you can get all sorts of relative paths and edge cases.